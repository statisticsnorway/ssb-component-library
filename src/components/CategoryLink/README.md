# Category link

### Usage

#### HTML

```html
<a class="ssb-category-link" href=" ">
  <div class="div-wrapper">
    <div class="text-divs title-wrapper">
      <span class="link-title-text">Category link with description</span>
    </div>
    <div class="text-divs sub-wrapper">
      <span class="link-sub-text">Description goes here</span>
    </div>
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width="24"
      height="24"
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      stroke-width="2"
      stroke-linecap="round"
      stroke-linejoin="round"
      class="arrow-icon"
    >
      <line x1="5" y1="12" x2="19" y2="12"></line>
      <polyline points="12 5 19 12 12 19"></polyline>
    </svg>
  </div>
</a>
```

#### React

```jsx harmony
<CategoryLink href='' titleText='Category link with description' subText='Description goes here' />
```

Available props:

| Name      | Type            | Description                                                   |
| --------- | --------------- | ------------------------------------------------------------- |
| ariaLabel | string          | Defines a string value that describe the link                 |
| className | string          | Optional container class                                      |
| href      | Required string | Destination for navigation                                    |
| titleText | Required string | text string for the title of the link                         |
| subText   | string          | text string below the title to describe the link              |
| tabIndex  | number          | Tab index for focus                                           |
| title     | string          | The title attribute is used to provide additional information |
