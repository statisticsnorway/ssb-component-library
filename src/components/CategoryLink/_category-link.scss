$default-font-size: 16px;
$profiled-font-size: 18px;
$header-font-size: 20px;
$icon-padding: 8px;

.ssb-category-link {
  @include open-sans;
  align-items: center;
  background-color: $ssb-white;
  background-image: linear-gradient(120deg, $ssb-green-4 0%, $ssb-green-4 100%);
  background-position: 0 100%;
  background-repeat: no-repeat;
  background-size: 100% 0;
  border: 1px solid $ssb-dark-2;
  border-bottom: 3px solid $ssb-green-4;
  color: $ssb-green-4;
  cursor: pointer;
  display: block;
  font-size: $default-font-size;
  padding: 20px 40px;
  line-height: 1.7;
  position: relative;
  text-decoration: none;
  transition:
    background-size 0.2s ease-in,
    color 0.1s;
  width: 100%;
  box-sizing: border-box;

  @media #{$mobile} {
    padding: 20px;
  }

  .div-wrapper::after {
    content: '';
    display: table;
    clear: both;
  }

  .div-wrapper {
    display: grid;
    grid-template-columns: auto 24px;
    grid-template-rows: auto auto;
  }

  .icon-wrapper {
    grid-row: 1 / 3;
    grid-column: 2;
    align-self: center;
    align-content: center;
  }

  .title-wrapper {
    grid-column: 1;
    grid-row: 1;
  }

  .sub-wrapper {
    grid-column: 1;
    grid-row: 2;
    margin-top: 3px;
  }

  .link-title-text {
    @include roboto;
    color: $ssb-dark-6;
    font-size: $header-font-size;
    font-weight: bold;
  }

  .link-sub-text {
    color: $ssb-dark-6;
  }

  &:hover,
  &:focus {
    background-size: 100% 100%;
    color: $ssb-white;
    outline: 0;
    text-decoration: none;

    .link-title-text,
    .link-sub-text {
      color: $ssb-white;
    }
  }

  .link-text {
    color: $ssb-green-4;
    transition: color 0.1s;

    &:hover,
    &:focus {
      color: $ssb-white;
      outline: 0;
    }
  }
}
