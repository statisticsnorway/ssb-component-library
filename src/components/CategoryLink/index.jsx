import React from 'react'
import { ArrowRight } from 'react-feather'
import PropTypes from 'prop-types'

const CategoryLink = ({ ariaLabel, titleText, subText, className, href, tabIndex, title }) => {
  const classNames = `ssb-category-link${className ? ` ${className}` : ''}`

  return (
    <a className={classNames} href={href} tabIndex={tabIndex} aria-label={ariaLabel} title={title}>
      {titleText && (
        <div className='div-wrapper'>
          <div className='text-divs title-wrapper'>
            <span className='link-title-text'>{titleText}</span>
          </div>
          <div className='text-divs sub-wrapper'>
            <span className='link-sub-text'>{subText}</span>
          </div>
          <div className='icon-wrapper'>
            <ArrowRight className='arrow-icon' size={24} />
          </div>
        </div>
      )}
    </a>
  )
}

CategoryLink.defaultProps = {
  className: '',
}

CategoryLink.propTypes = {
  ariaLabel: PropTypes.string,
  titleText: PropTypes.string,
  subText: PropTypes.string,
  className: PropTypes.string,
  href: PropTypes.string.isRequired,
  tabIndex: PropTypes.number,
  title: PropTypes.string,
}

export default CategoryLink
