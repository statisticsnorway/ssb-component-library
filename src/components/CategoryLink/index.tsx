import React from 'react'
import { ArrowRight } from 'react-feather'
import classNames from '../../utils/utils'

interface CategoryLinkProps {
  ariaLabel?: string
  titleText?: string
  subText?: string
  className?: string
  href: string
  tabIndex?: number
  title?: string
}

const CategoryLink: React.FC<CategoryLinkProps> = ({
  ariaLabel,
  titleText,
  subText,
  className,
  href,
  tabIndex,
  title,
}) => (
  <a
    className={classNames('ssb-category-link', className)}
    href={href}
    tabIndex={tabIndex}
    aria-label={ariaLabel}
    title={title}
  >
    {titleText && (
      <div className='div-wrapper'>
        <div className='text-divs title-wrapper'>
          <span className='link-title-text'>{titleText}</span>
        </div>
        <div className='text-divs sub-wrapper'>
          <span className='link-sub-text'>{subText}</span>
        </div>
        <div className='icon-wrapper'>
          <ArrowRight className='arrow-icon' size={24} />
        </div>
      </div>
    )}
  </a>
)

export default CategoryLink
