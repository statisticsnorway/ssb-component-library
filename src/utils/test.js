import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
