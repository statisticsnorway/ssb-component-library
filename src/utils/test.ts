import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
