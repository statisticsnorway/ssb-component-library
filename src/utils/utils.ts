// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
