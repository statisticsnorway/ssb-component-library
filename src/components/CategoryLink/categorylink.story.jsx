import React from 'react'
import CategoryLink from './index'

export default {
  title: 'CategoryLink',
}

export const Default = () => (
  <div style={{ width: '1200px', display: 'flex', justifyContent: 'center' }}>
    <div style={{ width: '580px', marginRight: '20px' }}>
      <CategoryLink href=' ' titleText='Category link with description' subText='Description goes here' />
    </div>
    <div style={{ width: '580px' }}>
      <CategoryLink href=' ' titleText='Category link with description' subText='Description goes here' />
    </div>
  </div>
)
