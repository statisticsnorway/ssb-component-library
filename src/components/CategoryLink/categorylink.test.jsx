import React from 'react'
import { render } from '../../utils/test'
import CategoryLink from './index'

describe('CategoryLink component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(
      <CategoryLink href='' titleText='Category link with description' subText='Description goes here' />
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
