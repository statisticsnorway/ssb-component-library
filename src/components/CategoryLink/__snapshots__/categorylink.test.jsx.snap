// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryLink component Matches the snapshot 1`] = `
<DocumentFragment>
  <a
    class="ssb-category-link"
    href=""
  >
    <div
      class="div-wrapper"
    >
      <div
        class="text-divs title-wrapper"
      >
        <span
          class="link-title-text"
        >
          Category link with description
        </span>
      </div>
      <div
        class="text-divs sub-wrapper"
      >
        <span
          class="link-sub-text"
        >
          Description goes here
        </span>
      </div>
      <div
        class="icon-wrapper"
      >
        <svg
          class="arrow-icon"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="5"
            x2="19"
            y1="12"
            y2="12"
          />
          <polyline
            points="12 5 19 12 12 19"
          />
        </svg>
      </div>
    </div>
  </a>
</DocumentFragment>
`;
